packages/medusa/src/services/product.ts

```diff
@@ -0,0 +1,1088 @@
+  protected readonly imageRepository_: typeof ImageRepository
+  protected readonly cartRepository_: typeof CartRepository
+  protected readonly productVariantService_: ProductVariantService
+  protected readonly productCollectionService_: ProductCollectionService
+  protected readonly searchService_: SearchService
+  protected readonly eventBus_: EventBusService
+  protected readonly priceSelectionStrategy_: PriceSelectionStrategy
+
+  static readonly IndexName = `products`
+  static readonly Events = {
+    UPDATED: "product.updated",
+    CREATED: "product.created",
+    DELETED: "product.deleted",
+  }
+
+  constructor({
+    manager,
+    productRepository,
+    productVariantRepository,
+    productOptionRepository,
+    eventBusService,
+    productVariantService,
+    productCollectionService,
+    productTypeRepository,
+    productTagRepository,
+    imageRepository,
+    searchService,
+    cartRepository,
+    priceSelectionStrategy,
+  }: InjectedDependencies) {
+    super({
+      manager,
+      productRepository,
+      productVariantRepository,
+      productOptionRepository,
+      eventBusService,
+      productVariantService,
+      productCollectionService,
+      productTypeRepository,
+      productTagRepository,
+      imageRepository,
+      searchService,
+      cartRepository,
+      priceSelectionStrategy,
+    })
+
+    this.manager_ = manager
+
+    this.productOptionRepository_ = productOptionRepository
+
+    this.productRepository_ = productRepository
+
+    this.productVariantRepository_ = productVariantRepository
+
+    this.eventBus_ = eventBusService
+
+    this.productVariantService_ = productVariantService
+
+    this.productCollectionService_ = productCollectionService
+
+    this.productTypeRepository_ = productTypeRepository
+
+    this.productTagRepository_ = productTagRepository
+
+    this.imageRepository_ = imageRepository
+
+    this.searchService_ = searchService
+
+    this.cartRepository_ = cartRepository
+
+    this.priceSelectionStrategy_ = priceSelectionStrategy
+  }
+
+  /**
+   * Lists products based on the provided parameters.
+   * @param selector - an object that defines rules to filter products
+   *   by
+   * @param config - object that defines the scope for what should be
+   *   returned
+   * @return the result of the find operation
+   */
+  async list(
+    selector: FilterableProductProps | Selector<Product> = {},
+    config: FindProductConfig = {
+      relations: [],
+      skip: 0,
+      take: 20,
+      include_discount_prices: false,
+    }
+  ): Promise<Product[]> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+
+      const priceIndex = config.relations?.indexOf("variants.prices") ?? -1
+      if (priceIndex >= 0 && config.relations) {
+        config.relations = [...config.relations]
+        config.relations.splice(priceIndex, 1)
+      }
```

Comment:
suggestion, very nit(non-blocking): We are removing variant prices as a relation here, because we are now letting the price selection strategy set additional prices. I think a short comment on that here, could be useful for a reader without that context. Wdty?

Update: See other comment on extracting this to a util. With proper naming, we could get two birds with one stone.

Reply:
Very true! I made some changes, let me know what you think 😄

Reply:
Think this is a very neat solution! 🙌